engine/source_sym_engine.ml

```diff
@@ -0,0 +1,175 @@
+
+type hashable_constructor = | HConstructor of constructor | HWildcard
+
+let sym_exec source =
+  let group_constructors rows : (constructor * row list) list * row list =
+    let grouptbl = Hashtbl.create 42
+    in
+    let (ksttbl: (source_blackbox , int) Hashtbl.t) = Hashtbl.create 42
+    in
+    let narity_of_k = function
+      | Nil -> 0
+      | Int _ | Bool _ | String _ -> 1
+      | Cons -> 2
+      | Tuple n -> n
+      | Variant v -> Hashtbl.find ksttbl v
+    in
+    let nary_wildcard n wildcards =
+      let rec inc_narity n ((wlist, expr):pattern list * source_expr) =
+        assert ( n >= 0);
+        if n > 0
+        then inc_narity (n-1) (Wildcard::wlist, expr)
+        else (wlist, expr)
+      in
+      List.map (inc_narity (n-List.length wildcards)) wildcards
+    in
+    let rec first_scan : pattern list -> unit = function
+      | [] -> ()
+      | (pattern::ptl) ->
+        match pattern with
+        | Wildcard -> ()
+        | Or (p1, p2) -> first_scan [p1; p2]
+        | As (p, _) -> first_scan [p]
+        | Constructor (k, plist) ->
+          match k with
+          | Variant v -> 
+            let len_ptl = List.length plist in
+            begin match Hashtbl.find_opt ksttbl v with 
+              | Some (narity:int) -> if len_ptl > narity then Hashtbl.replace ksttbl v len_ptl
+              | None -> Hashtbl.add ksttbl v len_ptl
+            end;
+            first_scan plist;
+            first_scan ptl
+          | _ -> () (* Don't consider constructors that are not variants *)
+    in
+    let rec put_in_group : pattern list * source_expr -> unit = function
+      | ([], _expr) -> assert false
+      | ((pattern::ptl), expr) ->
+        match pattern with
+        | Constructor (k, plist) ->
+          let binding = match Hashtbl.find_opt grouptbl (HConstructor k) with
+            | Some (idx, lst) -> (idx, (plist@ptl, expr)::lst)
+            | None -> let idx = Hashtbl.length grouptbl in
+              (idx, (plist@ptl, expr)::[])
+          in
+          Hashtbl.replace grouptbl (HConstructor k) binding
+        | Wildcard ->
+          let binding = match Hashtbl.find_opt grouptbl HWildcard with
+          | Some (idx, lst) -> (idx, (ptl, expr)::lst)
+          | None -> let idx = Hashtbl.length grouptbl in
+            (idx, (ptl, expr)::[])
+          in
+          Hashtbl.replace grouptbl HWildcard binding
+        | As (pattern, _) -> put_in_group (pattern::ptl, expr)
+        | Or (p1, p2) -> put_in_group (p1::ptl, expr); put_in_group (p2::ptl, expr) 
+    in
+    List.iter (fun row -> fst row |> first_scan) rows;
+    List.iter put_in_group rows;
+    let rec take_until_wildcards accum wildcards = function
+      | [] -> accum, wildcards
+      | hd::tl -> let (hk, (idx, rows)) = hd in
+        match hk with
+        | HWildcard ->  assert (wildcards = (0, [])); take_until_wildcards accum (idx, rows) tl
+        | HConstructor k -> take_until_wildcards ((k, rows)::accum) wildcards tl
+    in
```

Comment:
The `take_until_wildcards` business looks complex, I don't understand the details and I'm not sure it is correct. The approach I had in mind was to use the following logic in `put_in_group`:
- if the head pattern is a constructor, put it in the constructor group
- if the head pattern is a wildcard, put it in the wildcard group *and* in all the constructor groups in turn

Doing things this way, at the end of this traversal you have all the groups with rows in the right (reverse) order.
This assumes that you know what "all the constructor groups" are at any point (and not just: a group for each constructor that I have already encountered so far in my second scan), which requires that the first scan collects all constructors, not just variants.